Add Gallery Block

mkaz opened the pull request:
Initial PR for adding the gallery block - the initial code just stubs out the basics. Will need to add in multiple style galleries, drag-drop reordering, image upload, and any additional styles, will break these into additional smaller issues so it's not all done under one large PR.

See Issue #317

mkaz committed:
Initial gallery block

Creates initial gallery block, with stubbed out functionality.
Will create tickets for drag/drop, upload, multiple styles

mkaz committed:
Add gallery to library list

jasmussen commented:
Thrilled you're working on this! I tried building but it's erroring for me. But I see that it's in progress, let me know if/when you'd like eyes on this.

mkaz committed:
Fix paths for imports

mkaz committed:
Make the linter happy

jasmussen commented:
Nice first stab! 

Can you use the new Placeholder components instead of making the placeholder manually? See also embed.js for an example. I know the design is slightly different.

It would be really good to hook up the media library, I think lots of pieces fall in place then. 

🎉

mkaz committed:
Switch out with Placeholder component

mkaz committed:
Integrate Media Library button into Gallery

mkaz commented:
@jasmussen Media Library integrated, check it out. Any suggestions on removing images already added or adding more? We could click the whole block and take back into Media Library?

westonruter commented:
Cross reference core Gallery widget: https://github.com/xwp/wp-core-media-widgets/pull/120

jasmussen commented:
> Media Library integrated, check it out. Any suggestions on removing images already added or adding more? We could click the whole block and take back into Media Library?

I tweaked/updated some mockups to explore this better.

Neutral:

![gallery neutral](https://cloud.githubusercontent.com/assets/1204802/26348010/9aa207ce-3fab-11e7-9872-475382378c2d.png)

Selected:

![gallery selected](https://cloud.githubusercontent.com/assets/1204802/26348020/9d9d7634-3fab-11e7-8975-9a0450d8ad23.png)

To explain the blueprint —  the toolbar that shows up on the block itself is the "Quick Toolbar". The "Edit" button on that toolbar, same as for the gallery block in WordPress currently, opens the media library on the "Edit Gallery" page, that looks like this:

![screen shot 2017-05-23 at 11 34 25](https://cloud.githubusercontent.com/assets/1204802/26348076/d1180baa-3fab-11e7-9111-82ae5ddac903.png)

The sidebar on the right, is the "Inspector". The intricate details here are still a little bit in flux, and subject to change, so these are best handled in a separate PR. But essentially these contain more advanced aspects of editing the block, than can fit in the quick toolbar.

klihelp commented:
looks good:)

mkaz commented:
I understand for the Quick Toolbar to include the Edit which will go back to Media Library, and the drop down to pick the different styles. However, I'm not sure what the alignment (left, right, center, and stretch) means for a gallery

jasmussen commented:
> However, I'm not sure what the alignment (left, right, center, and stretch) means for a gallery

So keep in mind we're building desktop publishing for the web. And so it makes sense to think of a gallery more as "a blob of images" than a traditional gallery, which is just a kitchen sink full of photos.

For example, imagine a gallery with 3 pictures, set to 1 column in the options, floated left, with text flowing around it on the right. 

Or imagine a full-width/stretch'ed gallery set to 3 columns, as a separator between two paragraphs. 

The alignments are basically "layout tools", and users can do anything with them they like. Though ideally we'd be able to re-use the alignments from the image block 1:1 without much in the way of changing it. 

Some quick and dirty mockups... align none:

![align none](https://cloud.githubusercontent.com/assets/1204802/26370842/5751e248-3ff9-11e7-9ef7-89cfe7515a15.png)

Align full width:

![gallery full width](https://cloud.githubusercontent.com/assets/1204802/26370844/5ae9e3a6-3ff9-11e7-9449-d1f413dc3a7b.png)

mkaz committed:
Add controls for alignment, and start to edit opening back media library

mkaz committed:
Change CSS to just blocks-gallery to fix image width issue

mkaz committed:
Switch on open to Edit Gallery with images

Instead of reopening the Media Gallery, which the Media Upload Button does,
when editing the gallery requires opening the Edit Gallery modal. Also
requires setting selection and event which translates backbone data.

Props to @georgeh, who I paired with developing

See wp.media.editor from wp-includes/js/media-views.js

nylen commented:
After #849 which added a bunch of tests for block parsing and serialization, this PR should probably be rebased against `master`, either because the build will fail upon merge, or because this PR can add some new tests using this structure ([details here](https://github.com/WordPress/gutenberg/blob/484c146da24c8dac8232452f559161d58023c5fd/blocks/test/fixtures/README.md)).

mkaz committed:
Fix query/set attributes

mkaz committed:
Initial fixture for testing

mkaz commented:
Hi @nylen - I've rebased and updated the block including the new text fixture.
I'm a little stuck trying to get it to generate the .json and .serialized files - could you help by taking a look.  
```
 1) full post content fixture core-gallery:
     TypeError: Cannot read property 'children' of undefined
```

I'm seeing the above error. Once the tests part get straightened out, the block should be ready to go - so can use a review too :-)

nylen committed:
Work around receiving objects that are not React components

nylen committed:
Add remaining `core/gallery` test fixtures

nylen commented:
@mkaz the issues with the test framework itself are addressed in the latest few commits.  However, these tests are pointing to some problems with the block.  The markup in the `core-gallery.html` fixture does not have the format expected by the `attributes` matchers:

```js
url: attr( 'data-url' ),
alt: attr( 'data-alt' ),
```

As a result, these attributes are passed through the block as `{ url: undefined, alt: undefined }`, which is converted to `{}` when serialized to JSON.  Here is the resulting JSON fixture file, which should contain details about the images instead:

```js
[
    {
        "uid": "_uid_0",
        "blockType": "core/gallery",
        "attributes": {
            "images": [
                {},
                {}
            ]
        }
    }
]
```

Previously, any object passing through that part of the test code was assumed to be a React component, which was causing the specific issue you were seeing.  I've fixed that.

Separately, I don't think we should duplicate the URL and `alt` in the `data-url` and `data-alt` attributes at all.  Why not just grab them straight from the `img` tag?

mkaz commented:
Thanks @nylen - I thought I saw somewhere it was preferred to include the attributes in data, I'll switch to get them out of the img tag itself

mkaz committed:
Switch attributes to use img attributes and not add data attributes

mkaz committed:
Fix to registerBlockType

mtias commented on blocks/library/gallery/gallery-image.js:
```diff
@@ -0,0 +1,10 @@
+
+export default class GalleryImage extends wp.element.Component {
```
Dos this need to be an `element.Component` or could it be just a function?

mtias commented on blocks/test/fixtures/core-gallery.html:
```diff
@@ -0,0 +1,10 @@
+<!-- wp:core/gallery -->
+<div class="blocks-gallery">
+	<div class="blocks-gallery-image">
```
These could be `figure` instead of `div` for the markup.

aduth commented on blocks/library/gallery/gallery-image.js:
```diff
@@ -0,0 +1,10 @@
+export default class GalleryImage extends wp.element.Component {
+	render() {
+		return (
+			<div key={ this.props.i } className="blocks-gallery-image">
```
`key` here isn't doing anything.

aduth replied:
This wasn't addressed.

aduth commented on blocks/library/gallery/index.js:
```diff
@@ -0,0 +1,142 @@
+					<MediaUploadButton
+						onSelect={ setMediaUrl }
+						type="image"
+						auto-open
```
Curious why we chose dash-separated here, instead of `autoOpen`. Camel-case to be the convention with prop naming.

mkaz replied:
This is how it was specified in MediaUploadButton, I'd prefer to ship and then fix in another PR since it touches a couple blocks at once. Created issue #1083 to handle

aduth commented on blocks/library/gallery/index.js:
```diff
@@ -0,0 +1,142 @@
+		} );
+	}
+
+	editFrame.on( 'insert', updateFn );
```
Do we need to detach these event handlers after the edit frame is closed, to avoid risk of zombie event listeners?

mkaz replied:
On the frame Unmount the entire frame is destroyed, so the event handlers should be taken care of with that. See https://github.com/WordPress/gutenberg/blob/master/blocks/media-upload-button/index.js#L36

mkaz committed:
Update test to use <figure> instead of <div> for images

mkaz committed:
Switch attributes to use figure for querying

mkaz committed:
Switch to simple function, use <figure> instead of <div>

mkaz commented:
Thanks @mtias updated with your suggestions and switched out to simple function instead of component :+1:

mkaz committed:
Merge branch 'master' into add/317-gallery-block

mtias commented:
Thanks! Let's get it in and iterate.

jasmussen commented:
🎉

StaggerLeee commented:
Would like to ask for one non-GUI option for Admins, filter. To be able to limit max number of Items per gallery ?
Actually two in one, to be able to limit max number of galleries per Page/Post.

What is your advice, how can we continue to call "filters" now in React framework ?

mkaz commented:
@StaggerLeee Are you looking for an option to limit a user on how many items they can put into a gallery? Is this necessary?

I think something like that could easily be circumnavigated by inserting two gallery blocks into the same post, or just inserting multiple images. 

Can you create a new issue for this discussion?

StaggerLeee commented:
Maybe in some later stage. Now you all have to much to do.

I meant you moved everything to blocks, all widgets. When they are used on Homepage template could be nice to have some limit option, and still give User power to change things on Homepage.
